crates/sequencer/src/sequencer.rs

```diff
@@ -780,22 +676,159 @@ where
+                // The RPC from which the sender can be called is only registered for test mode. This means
+                // that evey though we check the receiver here, it'll never be "ready" to be consumed unless in test mode.
+                _ = self.l2_force_block_rx.next(), if self.config.test_mode => {
+                    if missed_da_blocks_count > 0 {
+                        debug!("We have {} missed DA blocks", missed_da_blocks_count);
+                        for _ in 1..=missed_da_blocks_count {
+                            let needed_da_block_height = last_used_l1_height + 1;
+                            let da_block = self
+                                .da_service
+                                .get_block_at(needed_da_block_height)
+                                .await
+                                .map_err(|e| anyhow!(e))?;
+
+                            debug!("Created an empty L2 for L1={}", needed_da_block_height);
+                            if let Err(e) = self.produce_l2_block(da_block, l1_fee_rate, L2BlockMode::Empty, &pg_pool, &mut last_used_l1_height).await {
+                                error!("Sequencer error: {}", e);
+                            }
+                        }
+                        missed_da_blocks_count = 0;
+                    }
+
+                    if let Err(e) = self.produce_l2_block(last_finalized_block.clone(), l1_fee_rate, L2BlockMode::NotEmpty, &pg_pool, &mut last_used_l1_height).await {
+                        error!("Sequencer error: {}", e);
+                    }
+                },
+                // If sequencer is in production mode, it will build a block every 2 seconds
+                _ = interval.tick(), if !self.config.test_mode => {
+                    // By default, we produce a non-empty block IFF we were caught up all the way to
+                    // last_finalized_block. If there are missed DA blocks, we start producing
+                    // empty blocks at ~2 second rate, 1 L2 block per respective missed DA block
+                    // until we know we caught up with L1.
+                    let da_block = last_finalized_block.clone();
+
+                    if missed_da_blocks_count > 0 {
+                        debug!("We have {} missed DA blocks", missed_da_blocks_count);
+                        for _ in 1..=missed_da_blocks_count {
+                            let needed_da_block_height = last_used_l1_height + 1;
+                            let da_block = self
+                                .da_service
+                                .get_block_at(needed_da_block_height)
+                                .await
+                                .map_err(|e| anyhow!(e))?;
+
+                            debug!("Created an empty L2 for L1={}", needed_da_block_height);
+                            if let Err(e) = self.produce_l2_block(da_block, l1_fee_rate, L2BlockMode::Empty, &pg_pool, &mut last_used_l1_height).await {
```

Comment:
i know i did this but is there a better way than passing mut ref to a function to change the last_used_l1_height?